airbyte-workers/src/main/java/io/airbyte/workers/DefaultReplicationWorker.java

```diff
@@ -52,28 +59,34 @@
   private final Source<AirbyteMessage> source;
   private final Mapper<AirbyteMessage> mapper;
   private final Destination<AirbyteMessage> destination;
-  private final MessageTracker<AirbyteMessage> messageTracker;
+  private final MessageTracker<AirbyteMessage> sourceMessageTracker;
+  private final MessageTracker<AirbyteMessage> destinationMessageTracker;
 
   private final AtomicBoolean cancelled;
+  private final AtomicBoolean hasFailed;
 
   public DefaultReplicationWorker(final String jobId,
                                   final int attempt,
                                   final Source<AirbyteMessage> source,
                                   final Mapper<AirbyteMessage> mapper,
                                   final Destination<AirbyteMessage> destination,
-                                  final MessageTracker<AirbyteMessage> messageTracker) {
+                                  final MessageTracker<AirbyteMessage> sourceMessageTracker,
+                                  final MessageTracker<AirbyteMessage> destinationMessageTracker) {
     this.jobId = jobId;
     this.attempt = attempt;
     this.source = source;
     this.mapper = mapper;
     this.destination = destination;
-    this.messageTracker = messageTracker;
+    this.sourceMessageTracker = sourceMessageTracker;
+    this.destinationMessageTracker = destinationMessageTracker;
 
     this.cancelled = new AtomicBoolean(false);
+    this.hasFailed = new AtomicBoolean(false);
   }
 
+  @SuppressWarnings("OptionalIsPresent")
   @Override
-  public StandardSyncOutput run(StandardSyncInput syncInput, Path jobRoot) throws WorkerException {
+  public ReplicationOutput run(StandardSyncInput syncInput, Path jobRoot) throws WorkerException {
```

Comment:
might just be my preference, since i'm the 3rd person to read this - what if we break this big block up for readability? it's somewhat difficult to parse now.

something like:
```
  @SuppressWarnings("OptionalIsPresent")
  @Override
  public ReplicationOutput run(StandardSyncInput syncInput, Path jobRoot) throws WorkerException {
    LOGGER.info("start sync worker. job id: {} attempt id: {}", jobId, attempt);

    // todo (cgardens) - this should not be happening in the worker. this is configuration information
    // that is independent of workflow executions.
    final StandardTargetConfig destinationConfig = WorkerUtils.syncToTargetConfig(syncInput);
    destinationConfig.setCatalog(mapper.mapCatalog(destinationConfig.getCatalog()));

    long startTime = System.currentTimeMillis();
    try {
      LOGGER.info("configured sync modes: {}", syncInput.getCatalog().getStreams()
          .stream()
          .collect(Collectors.toMap(s -> s.getStream().getNamespace() + "." + s.getStream().getName(),
              s -> String.format("%s - %s", s.getSyncMode(), s.getDestinationSyncMode()))));
      final StandardTapConfig sourceConfig = WorkerUtils.syncToTapConfig(syncInput);

      // note: resources are closed in the opposite order in which they are declared. thus source will be
      // closed first (which is what we want).
      final ExecutorService executorService = Executors.newFixedThreadPool(2);
      try (destination; source) {
        destination.start(destinationConfig, jobRoot);
        source.start(sourceConfig, jobRoot);

        final Future<?> destinationThreadExecution = executorService.submit(readDstUntilDoneOrCancelled());
        final Future<?> sourceThreadExecution = executorService.submit(readSrcUntilDoneOrCancelled());

        LOGGER.info("Waiting for source thread to join.");
        sourceThreadExecution.get();
        LOGGER.info("Source thread complete.");
        LOGGER.info("Waiting for destination thread to join.");
        destinationThreadExecution.get();
        LOGGER.info("Destination thread complete.");

      } catch (Exception e) {
        hasFailed.set(true);
        LOGGER.error("Sync worker failed.", e);
      } finally {
        executorService.shutdownNow();
      }

      final ReplicationStatus outputStatus =
          cancelled.get() ? ReplicationStatus.CANCELLED : hasFailed.get() ? ReplicationStatus.FAILED : ReplicationStatus.COMPLETED;
      final ReplicationAttemptSummary summary = new ReplicationAttemptSummary()
          .withStatus(outputStatus)
          .withRecordsSynced(sourceMessageTracker.getRecordCount())
          .withBytesSynced(sourceMessageTracker.getBytesCount())
          .withStartTime(startTime)
          .withEndTime(System.currentTimeMillis());

      LOGGER.info("sync summary: {}", summary);

      final ReplicationOutput output = new ReplicationOutput()
          .withReplicationAttemptSummary(summary)
          .withOutputCatalog(destinationConfig.getCatalog());

      if (destinationMessageTracker.getOutputState().isPresent()) {
        LOGGER.info("State capture: Updated state to: {}", destinationMessageTracker.getOutputState());
        final State state = new State().withState(destinationMessageTracker.getOutputState().get());
        output.withState(state);
      } else if (syncInput.getState() != null) {
        LOGGER.warn("State capture: No new state, falling back on input state: {}", syncInput.getState());
        output.withState(syncInput.getState());
      } else {
        LOGGER.warn("State capture: No state retained.");
      }

      return output;
    } catch (Exception e) {
      throw new WorkerException("Sync failed", e);
    }

  }

  private Thread readSrcUntilDoneOrCancelled() {
    return new Thread(() -> {
      try {
        while (!cancelled.get() && !source.isFinished()) {
          final Optional<AirbyteMessage> messageOptional = source.attemptRead();
          if (messageOptional.isPresent()) {
            final AirbyteMessage message = mapper.mapMessage(messageOptional.get());

            LOGGER.info("record in DefaultReplicationWorker: {}", message);

            sourceMessageTracker.accept(message);
            destination.accept(message);
          }
        }
        destination.notifyEndOfStream();
      } catch (Exception e) {
        throw new RuntimeException(e);
      }
    });
  }

  private Thread readDstUntilDoneOrCancelled() {
    return new Thread(() -> {
      try {
        while (!cancelled.get() && !destination.isFinished()) {
          final Optional<AirbyteMessage> messageOptional = destination.attemptRead();
          if (messageOptional.isPresent()) {
            LOGGER.info("state in DefaultReplicationWorker from Destination: {}", messageOptional.get());
            destinationMessageTracker.accept(messageOptional.get());
          }
        }
      } catch (Exception e) {
        throw new RuntimeException(e);
      }
    });
  }
```
main changes are pulling out the thread set up + bringing the executor creation to where it's first used to make it more obvious what the main loop is really running & get rid of the nested exceptions (functionally identical exception handling but this is easier to read).

Reply:
updated. let me know if this helps.

Reply:
Nice! Those are better names than mine hah